### PR TITLE
Handle not enough keys from failed preconditions errors

### PR DESCRIPTION
**Pull request by narape**
# Hey, I just made a Pull Request!

## Description
Handle not enough keys from failed preconditions errors

## Motivation and Context
Google now returns a different error to signal that it cannot create new keys. The code can handle both types of error now

This is the error produced
```
POST https://iam.googleapis.com/v1/projects/-/serviceAccounts/foo@bar.iam.gserviceaccount.com/keys
{
  "code": 400,
  "errors": [
    {
      "domain": "global",
      "message": "Precondition check failed.",
      "reason": "failedPrecondition"
    }
  ],
  "message": "Precondition check failed.",
  "status": "FAILED_PRECONDITION"
}
   at com.spotify.styx.docker.KubernetesGCPServiceAccountSecretManager.ensureServiceAccountKeySecret(KubernetesGCPServiceAccountSecretManager.java:128)
   at com.spotify.styx.docker.KubernetesDockerRunner.lambda$ensureSecrets$0(KubernetesDockerRunner.java:263)
   at java.base/java.util.Optional.map(Optional.java:265)
   at com.spotify.styx.docker.KubernetesDockerRunner.ensureSecrets(KubernetesDockerRunner.java:261)
   at com.spotify.styx.docker.KubernetesDockerRunner.start(KubernetesDockerRunner.java:217)
   at com.spotify.styx.docker.RoutingDockerRunner.start(RoutingDockerRunner.java:42)
   at jdk.internal.reflect.GeneratedMethodAccessor116.invoke(Unknown Source)
   at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
   at java.base/java.lang.reflect.Method.invoke(Method.java:566)
   at com.spotify.styx.monitoring.TracingProxy.invoke(TracingProxy.java:67)
   at com.sun.proxy.$Proxy61.start(Unknown Source)
   at jdk.internal.reflect.GeneratedMethodAccessor116.invoke(Unknown Source)
   at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
   at java.base/java.lang.reflect.Method.invoke(Method.java:566)
   at com.spotify.styx.monitoring.MeteredProxy.lambda$invoke$0(MeteredProxy.java:63)
   at javaslang.control.Try.of(Try.java:39)
   at com.spotify.styx.monitoring.MeteredProxy.invoke(MeteredProxy.java:61)
   at com.sun.proxy.$Proxy61.start(Unknown Source)
   at com.spotify.styx.state.handlers.DockerRunnerHandler.safeTransitionInto(DockerRunnerHandler.java:65)
   at com.spotify.styx.state.handlers.AbstractRunnerHandler.transitionInto(AbstractRunnerHandler.java:71)
   at jdk.internal.reflect.GeneratedMethodAccessor92.invoke(Unknown Source)
   at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
   at java.base/java.lang.reflect.Method.invoke(Method.java:566)
   at com.spotify.styx.monitoring.TracingProxy.invoke(TracingProxy.java:67)
   at com.sun.proxy.$Proxy62.transitionInto(Unknown Source)
   at com.spotify.styx.state.OutputHandler.lambda$fanOutput$0(OutputHandler.java:51)
   at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
   at com.spotify.styx.state.OutputHandler.lambda$fanOutput$1(OutputHandler.java:50)
   at com.spotify.styx.state.MDCDecoratingHandler.transitionInto(MDCDecoratingHandler.java:49)
   at com.spotify.styx.state.PersistentStateManager.postTransition(PersistentStateManager.java:356)
   at com.spotify.styx.state.PersistentStateManager.receive(PersistentStateManager.java:214)
   at com.spotify.styx.state.handlers.ExecutionDescriptionHandler.transitionInto(ExecutionDescriptionHandler.java:92)
   at jdk.internal.reflect.GeneratedMethodAccessor92.invoke(Unknown Source)
   at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
   at java.base/java.lang.reflect.Method.invoke(Method.java:566)
   at com.spotify.styx.monitoring.TracingProxy.invoke(TracingProxy.java:67)
   at com.sun.proxy.$Proxy62.transitionInto(Unknown Source)
   at com.spotify.styx.state.OutputHandler.lambda$fanOutput$0(OutputHandler.java:51)
   at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
   at com.spotify.styx.state.OutputHandler.lambda$fanOutput$1(OutputHandler.java:50)
   at com.spotify.styx.state.MDCDecoratingHandler.transitionInto(MDCDecoratingHandler.java:49)
   at com.spotify.styx.state.PersistentStateManager.postTransition(PersistentStateManager.java:356)
   at com.spotify.styx.state.PersistentStateManager.receive(PersistentStateManager.java:214)
   at com.spotify.styx.state.
EventRouter.receiveIgnoreClosed(EventRouter.java:72)
   at jdk.internal.reflect.GeneratedMethodAccessor90.invoke(Unknown Source)
   at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
   at java.base/java.lang.reflect.Method.invoke(Method.java:566)
   at com.spotify.styx.monitoring.TracingProxy.invoke(TracingProxy.java:67)
   at com.sun.proxy.$Proxy65.receiveIgnoreClosed(Unknown Source)
   at com.spotify.styx.Scheduler.sendDequeue(Scheduler.java:419)
   at com.spotify.styx.Scheduler.processInstance(Scheduler.java:319)
   at com.spotify.styx.Scheduler.lambda$processInstances$3(Scheduler.java:226)
   at io.opencensus.trace.CurrentSpanUtils$RunnableInSpan.run(CurrentSpanUtils.java:121)
   at io.opencensus.trace.SpanBuilder.startSpanAndRun(SpanBuilder.java:282)
   at com.spotify.styx.Scheduler.lambda$processInstances$4(Scheduler.java:224)
   at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1736)
   at io.grpc.Context$1.run(Context.java:579)
   at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
   at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
   at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
   at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
   at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
   at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
   Caused by: java.util.concurrent.ExecutionException: com.google.api.client.googleapis.json.GoogleJsonResponseException: 400 Bad Request
```

## Have you tested this? If so, how?
I have included unit tests.

## Checklist for PR author(s)

- [ ] Changes are covered by unit test
- [ ] All tests pass
- [ ] Code coverage check passes
- [ ] Error handling is tested
- [ ] Errors are handled at the appropriate layer
- [ ] Errors that cannot be handled where they occur are propagated
- [ ] (optional) Changes are covered by system test
- [ ] Relevant documentation updated
- [ ] This PR has NO breaking change to public API
- [ ] This PR has breaking change to public API and it is documented

## Checklist for PR reviewer(s)

- [ ] This PR has been incorporated in release note for the coming version
- [ ] Risky changes introduced by this PR have been all considered

### CODE REVIEW

File: styx-service-common/src/main/java/com/spotify/styx/util/GcpUtil.java [hunk 1]
```diff
@@ -23,9 +23,14 @@
 import com.google.api.client.googleapis.json.GoogleJsonError;
 import com.google.api.client.googleapis.json.GoogleJsonResponseException;
 import java.util.Optional;
+import java.util.function.Predicate;
 
 public class GcpUtil {
 
+  private static final int BAD_REQUEST = 400;
```
Comment (honnix) [inline styx-service-common/src/main/java/com/spotify/styx/util/GcpUtil.java:30]: We can get these from Apollo for free. `com.spotify.apollo.Status`

File: styx-scheduler-service/src/test/java/com/spotify/styx/docker/KubernetesGCPServiceAccountSecretManagerTest.java [hunk 2]
```diff
@@ -429,6 +429,23 @@ public void shouldHandleTooManyKeysCreated() throws IOException {
...
+  @Test
+  public void shouldHandleTooManyKeysCreatedFromFailedPrecondition() throws IOException {
+    when(serviceAccountKeyManager.serviceAccountExists(anyString())).thenReturn(true);
+
+    final GoogleJsonResponseException resourceExhausted = new GoogleJsonResponseException(
+        new HttpResponseException.Builder(400, "Precondition check failed", new HttpHeaders()),
+        new GoogleJsonError().set("status", "FAILED_PRECONDITION"));
+
+    doThrow(resourceExhausted).when(serviceAccountKeyManager).createJsonKey(any());
+    doThrow(resourceExhausted).when(serviceAccountKeyManager).createP12Key(any());
```
Comment (honnix) [inline styx-scheduler-service/src/test/java/com/spotify/styx/docker/KubernetesGCPServiceAccountSecretManagerTest.java:441]: Interesting that mockito doesn't complain we are over stubbing here. I think if we fail to create json key, we will not go and create p12 key.

**Review by honnix — APPROVED** [review]
Minor comments. Otherwise LGTM.

File: styx-scheduler-service/src/test/java/com/spotify/styx/docker/KubernetesGCPServiceAccountSecretManagerTest.java (hunk 2 above)
Comment (narape) [reply styx-scheduler-service/src/test/java/com/spotify/styx/docker/KubernetesGCPServiceAccountSecretManagerTest.java:441]: :shrug: I just copy and modified the test from above

**Comment by narape** [comment]
@honnix @bergman There is a problem with the end to end test but most probably is not related to this PR. I see that it have been failing for the last handfulls of building in circleci. Can you take a look?

**Pull request merged**